Asset Manager Documentation (#35)
ove/ove-asset-services: packages/ove-asset-manager/README.md

# OVE Asset Manager Services

This collection of microservices provides end users with a simple way of managing and processing content to be displayed using OVE. 

## Asset Manager

The core service is the **Asset Manager**. This enables media and content to be uploaded to an **S3** compatible **object store** and records metadata about its processing in an **SQL database**. A full user friendly website is provided to interact with the Asset Manager. A full RESTful API is provided for advanced users.  

> ## Asset Services

Reviewer: The next subsection refers to 'Asset Processing Microservices': it might be helpful to always include the 'Processing' (i.e. refer to 'Asset Processing Services' throughout).

Author: fixed

> A variety of other Asset Services are provided to process assets into a form advantageous for display using OVE. These services are generally headless and will automatically register themselves with an Asset Manager via its API. They will then generally automatically process assets which require their services. For example the [**Image Tile Service**](https://github.com/ove/ove-asset-services/blob/master/packages/ove-service-imagetiles/README.MD) will create a Deep Zoom image (.dzi file) and corresponding tileset from each image asset. 

Reviewer: Change `.MD` -> `.md` so the link works:

```suggestion
A variety of other Asset Services are provided to process assets into a form advantageous for display using OVE. These services are generally headless and will automatically register themselves with an Asset Manager via its API. They will then generally automatically process assets which require their services. For example the [**Image Tile Service**](https://github.com/ove/ove-asset-services/blob/master/packages/ove-service-imagetiles/README.md) will create a Deep Zoom image (.dzi file) and corresponding tileset from each image asset. 
```

Reviewer: This doesn't seem to have been fixed: the link is still to `README.MD` rather than `README.md`

Author: updated to use relative paths.

## Scalability

> The system is intended to be scalable and many copies of the Asset Manager can be run - the EntityFramework and SQL database concurrency models ensure that the ``ProcessingState`` for each Asset is atomically updated despite running multiple instances. 

Reviewer: ```suggestion
The system is intended to be scalable and many copies of the Asset Manager can be run - the [Entity Framework](https://docs.microsoft.com/en-us/aspnet/entity-framework) and SQL database concurrency models ensure that the ``ProcessingState`` for each Asset is atomically updated despite running multiple instances. 
```

> The Asset Processing Microservices work on a pull based model - they will each independently regularly request further work. You are encouraged to run multiple copies of each Asset Processing Microservice: these should be configured to communicate with an instance of the Asset Manager. These services will then periodically check for Assets to process. Your database and object store may each scale independently. 

Reviewer: Do they poll for work at a regular interval (if so, what?) or just announce when they have finished a task?

Author: They poll regularly for new work at an interval configurable through the appsettings.json. I'll add doc for this.

## Storage Model

> Assets will be stored on the object store on a one bucket per `Project` model. Every Asset is assigned a ``GUID`` on creation. This is used to identify it in the SQL database and on the object store. Within the Project bucket each Asset is uploaded to the root of a folder named by this ``GUID``. The filename of the asset will be preserved where possible - file names must consist of `a-z`,`A-Z`, `()`, `-` or `_` and be of length less than `1024` characters. Asset processing services are free to place processed files and folders anywhere within the asset directory. 

Reviewer: Capitalize "Asset processing services" for consistency:

```suggestion
Assets will be stored on the object store on a one bucket per `Project` model. Every Asset is assigned a ``GUID`` on creation. This is used to identify it in the SQL database and on the object store. Within the Project bucket each Asset is uploaded to the root of a folder named by this ``GUID``. The filename of the asset will be preserved where possible - file names must consist of `a-z`,`A-Z`, `()`, `-` or `_` and be of length less than `1024` characters. Asset Processing Services are free to place processed files and folders anywhere within the asset directory. 
```

## Configuration

> Configuration of the service is achieved by modifying the ``appsettings.json`` file. These settings are automatically over written by **Environment Variables** (reflection of the JSON hierarcy should be acheived using ``__`` e.g. ``s3Client__Secret``). 

Reviewer: "over written" -> "overwritten"

```suggestion
Configuration of the service is achieved by modifying the ``appsettings.json`` file. These settings are automatically overwritten by **Environment Variables** (reflection of the JSON hierarcy should be acheived using ``__`` e.g. ``s3Client__Secret``). 
```

### Configuring S3

> In common with all S3 compatible object stores the following three properties are required and should be set as follows in ``appsettings.json`` or overridden by ``environment variables`` as discussed above.

Reviewer: ```suggestion
In common with all S3 compatible object stores the following three properties are required and should be set as follows in ``appsettings.json`` or overridden by **Environment Variables** as discussed above.
```

```  
"s3Client": {
    "AccessKey": "key",
    "Secret": "secret",
    "ServiceURL": "host"
  }
 ```
 
### Configuring Maria DB

> The [MariaDB](https://mariadb.org/) configuration should be set as follows in `appsettings.json` or overridden by ``Environment Variables`` as discussed above. 

Reviewer: ```suggestion
The [MariaDB](https://mariadb.org/) configuration should be set as follows in `appsettings.json` or overridden by **Environment Variables** as discussed above. 
```

```
  "MariaDB": { 
    "ConnectionString": "Server=SERVER;Port=3306;Database=DATABASE;User=USERNAME;Password=PASSWORD;", 
    "Version" :  "10.3.10"  
  }
```

## Using the Asset Manager

> Assets should be uploaded to the asset manager either through the web interface or submitting a HTTP `POST` request to the `/OVEAssetModelController/Create/` API. 

Reviewer: ```suggestion
Assets should be uploaded to the asset manager either through the web interface or submitting a HTTP `POST` request to the `/OVEAssetModelController/Create/` API route. 
```

> Asset URLs on the object store may be found via the `/OVEAssetModelController/GetAssetURL?project=X&file=Y` API or by ID on  `/OVEAssetModelController/GetAssetURLbyId?id=X`

Reviewer: ```suggestion
Asset URLs on the object store may be found via the `/OVEAssetModelController/GetAssetURL?project=X&file=Y` API route or by `ID` on  `/OVEAssetModelController/GetAssetURLbyId?id=X`
```

## Versioning

> The Asset Manager supports file versioning. The service will, unless otherwise specified, always provide the most recently uploaded/edited asset version when searching by Project and Asset Name. Previous versions may be found via the `/OVEAssetModelController/ListAssets/Project/` and `/OVEAssetModelController/ListAssets/Project/Name/` API's.

Reviewer: ```suggestion
The Asset Manager supports file versioning. The service will, unless otherwise specified, always provide the most recently uploaded/edited asset version when searching by Project and Asset Name. Previous versions may be found via the `/OVEAssetModelController/ListAssets/Project/` and `/OVEAssetModelController/ListAssets/Project/Name/` API routes.
```

## Processing state

> Asset Processing Services will update the ProcessingState of each asset using the ``/OVEAssetModelController/SetProcessingState/{id}/{state}/`` API. Assets in the unprocessed state (`0`) will be returned to processing microservices upon request. To reset the processing state of an asset and retry the processing you should use the ``/OVEAssetModelController/ResetProcessing/{id}`` of the Asset Manager API. A custom set of procesing states may be registered for every Asset Type. 

Reviewer: ```suggestion
Asset Processing Services will update the ProcessingState of each asset using the ``/OVEAssetModelController/SetProcessingState/{id}/{state}/`` API route. Assets in the unprocessed state (`0`) will be returned to processing microservices upon request. To reset the processing state of an asset and retry the processing you should use the ``/OVEAssetModelController/ResetProcessing/{id}`` route of the Asset Manager API. A custom set of processing states may be registered for every Asset Type. 
```

Since the Asset Manager has no knowledge of different asset types, how to process them or how to display them each asset processing microservice must register this knowledge with the Asset Manager. Each microservice should be configured with the URI of an asset manager with the `AssetManagerHostUrl` configuration property. These services will then register with the asset manager using the `/api/ServicesRegistry/Register` route to register a service description (see [here](https://github.com/ove/ove-asset-services/blob/master/packages/ove-asset-manager/src/OVE.Service.AssetManager/Domain/OVEService.cs)). 

```
{
  "name": "AService",
  "fileTypes": [
    ".abc",
    ".xyz"
  ],
  "viewIFrameUrl": "full uri",
  "processingStates": {
    "int": "name",
  }
}
```

The `fileTypes` are used for validation of uploads to the asset manager.

The `processingStates` are used to provide meaningful status updates to users.

The `viewIFrameUrl` is used to enable each service to provide a webpage for rendering an asset for display within the asset manager. This **must** include the string ``{id}`` which will be replaced with id of the asset. 

## Asset Metadata

> Optionally assets may have JSON metadata attached which can be updated via `GET` / `POST` on `/OVEAssetModelController/AssetMeta/{id}`. This is intended to be entered programmatically by asset processing services. 

Reviewer: ```suggestion
Optionally assets may have JSON metadata attached which can be updated via `GET` / `POST` on `/OVEAssetModelController/AssetMeta/{id}`. This is intended to be entered programmatically by Asset Processing Services. 
```

## Implementation 

The Asset Manager is implemented in [**C#**](https://github.com/dotnet/roslyn) and [**.net core 2.1**](https://blogs.msdn.microsoft.com/dotnet/2018/05/30/announcing-net-core-2-1/) and so runs cross platform. 

> [**OWIN** (Open Web Interface for .Net)](http://owin.org/) is used to decouple the web stack modules with **Dependency Injection** used throughout, precise configuration can be explored in `Program.cs` and `Startup.cs`. Module imported are completed via [**Nuget**](www.nuget.org) and are listed in the `.csproj` files. 

Reviewer: ```suggestion
[**OWIN** (Open Web Interface for .Net)](http://owin.org/) is used to decouple the web stack modules with **Dependency Injection** used throughout, precise configuration can be explored in `Program.cs` and `Startup.cs`. Module imports are completed via [**Nuget**](www.nuget.org) and are listed in the `.csproj` files. 
```

The cross platform lightweight [**Kestrel**](https://github.com/aspnet/KestrelHttpServer) HTTP server is used. 

> The opensource [**ASP.Net**](https://github.com/aspnet/AspNetCore) framework is used to create RESTful APIs, further the [**Asp.net MVC**](https://github.com/aspnet/Mvc) framework and is used to coordinate views, controllers and model state. Model state validation is enabled via a [**Validation Attributes**](https://docs.microsoft.com/en-us/aspnet/core/mvc/models/validation?view=aspnetcore-2.1) and is provided server-side and automatically provided client-side with [**JQuery**](https://jquery.com/). 

Reviewer: ```suggestion
The open source [**ASP.Net**](https://github.com/aspnet/AspNetCore) framework is used to create RESTful APIs, further the [**Asp.net MVC**](https://github.com/aspnet/Mvc) framework and is used to coordinate views, controllers and model state. Model state validation is enabled via a [**Validation Attributes**](https://docs.microsoft.com/en-us/aspnet/core/mvc/models/validation?view=aspnetcore-2.1) and is provided server-side and automatically provided client-side with [**JQuery**](https://jquery.com/). 
```

# OVE Asset Manager Services

The objective of this collection of microservices is to provide end users with a simple way of managing and processing content to be displayed using OVE. 

## Asset Manager

The core service is the **Asset Manager** which enables upload and storage of media and content as well as maintaining meta data about its processing. Content is stored on an S3 compatible **ObjectStore**. Metadata is stored in an **SQL database**. A full user friendly website is provided to interact with the Asset Manager. A full RESTful API is provided for advanced users.  

## Asset Services

> A variety of other Asset Services are provided to process assets into a form advantageous for display using OVE. These services are generally headless and will automatically register themselves with an Asset Manager via API. They will then generally automatically process assets which require their services. For example the *Image Tile Service* will create a Deep Zoom Tileset from each image asset. 

Author: improved DZI comment. keeping with image asset for clarity

## Scalability

The system is intended to be scalable and many copies of the Asset Manager can be run - the EntityFramework and SQL database concurrency models means Asset `ProcessingState` is atomically updated despite running multiple instances. 

> You are encouraged to run multiple copies of each Asset Processing Microservice, these should be registered with an instance of the Asset Manager and will then periodically check for Assets to process. Your database and Objectstore may each scale independently. 

Author: loadbalencers are of course encouraged :)  have changed to `object store`

> You are encouraged to run multiple copies of each Asset Processing Microservice, these should be registered with an instance of the Asset Manager and will then periodically check for Assets to process. Your database and Objectstore may each scale independently. 

Author: thanks, clarified to "These services will then" as its pull not push based

## Storage Model

> Assets will be stored on the object store on a one bucket per `Project` model. Every Asset is assigned a GUID on creation. This is used to identify it in the SQL database and on the object store. Within the bucket each Asset is uploaded to the root of a folder named by this ``GUID``. The filename of the asset will be preserved where possible. Asset processing services are free to place processed files and folders anywhere within the asset directory. 

Author: documented this

## Configuration

> Configuration of the service is achieved by modifying the ``Appsettings.json`` file. These settings are automatically over written by **Environmental Variables** (reflection of the json hierarcy should be acheived using ``__`` e.g. ``s3Client__Secret``). 

Author: fixed

> Configuration of the service is achieved by modifying the ``Appsettings.json`` file. These settings are automatically over written by **Environmental Variables** (reflection of the json hierarcy should be acheived using ``__`` e.g. ``s3Client__Secret``). 

Reviewer 2: Issue still remains in Image Tile Service README

### Configuring S3

In common with all S3 compatible object stores the following three properties are required and should be set as follows in ``appsettings.json`` or overridden in environmental variables as discussed above.

```  
"s3Client": {
    "AccessKey": "secret",
    "Secret": "secret",
    "ServiceURL": "host"
  }
 ```
 
### Configuring Maria DB

> The MariaDB configuration should be set as follows in ``appsettings.json` or overridden in environmental variables as discussed above. 

Reviewer 2: ```suggestion
The MariaDB configuration should be set as follows in ``appsettings.json`` or overridden in environmental variables as discussed above. 
```

```
  "MariaDB": { 
    "ConnectionString": "Server=SERVER;Port=3306;Database=DATABASE;User=USERNAME;Password=PASSWORD;", 
    "Version" :  "10.3.10"  
  }
```

## Using the Asset Manager

Assets should be uploaded to the asset manager either through the web interface or posting to the `/OVEAssetModelController/Create/` API. 

> Asset URL's on the objectstore may be found via the `/OVEAssetModelController/GetAssetURL?project=X&File=Y` API or by ID on  

Reviewer 3: Should be on the same line.

```suggestion
Asset URL's on the objectstore may be found via the `/OVEAssetModelController/GetAssetURL?project=X&File=Y` API or by ID on  `/OVEAssetModelController/GetAssetURLbId?id=X`
```

Author: Sure, thanks

`/OVEAssetModelController/GetAssetURLbId?id=X`

## Versioning

> The Asset Manager supports file versioning and will be default always provide the most recently uploaded/edited asset version when searching by Project and Asset Name. Previous versions may be found via the `/OVEAssetModelController/ListAssets/Project/` and `/OVEAssetModelController/ListAssets/Project/Name/` API's.

Author: modified in a slighty different form to preserve meaning

## Processing state

> Asset processing services will update the ProcessingState of each asset using the ``/OVEAssetModelController/SetProcessingState/{id}/{state}/`` API. Assets in the unprocessed state (`0`) will be returned to processing microservices upon request. To reset the processing state of an asset and retry the processing you should use the ``/OVEAssetModelController/ResetProcessing/{id}`` API. 

Author: Hi James, the services are pull based,
if the processing state is reset then it will be reprocessed when a processing microservice requests for further work. I'll clarify this model earlier in the documentation.

## Interaction of Processing Services and Asset Manager

Since the Assetmanager has no knowledge of different asset types, how to process them or how to display them each asset processing microservice must register this knowledge with the Asset Manager. Each microservice should be configured with the URI of an asset manager with the `AssetManagerHostUrl` configuration property. These services will then register with the asset manager using the `/api/ServicesRegistry/Register` route to register a service description (see [here](https://github.com/ove/ove-asset-services/blob/master/packages/ove-asset-manager/src/OVE.Service.AssetManager/Domain/OVEService.cs)). 

```
{
  "name": "AService",

>   "fileTypes": [

Author: the asset manager is trusting the user to denote the fileType by selecting that service when uploading the asset. The service is then validating this choice by the fileTypes it accepts as denoted by the file extensions

>   "fileTypes": [

Author: (@jamesscottbrown i can rename this if necessary but i'd rather do it in a seperate issue/PR to avoid merge hell).

    ".abc",
    ".xyz"
  ],
  "viewIFrameUrl": "full uri",
  "processingStates": {
    "int": "name",
  }
}
```

The `fileTypes` are used for validation of uploads to the asset manager.

The `processingStates` are used to provide meaningful status updates to users.

The `viewIFrameUrl` is used to enable each service to provide a webpage for rendering an asset for display within the asset manager. This **must** include the string ``{id}`` which will be replaced with id of theasset. 

## Asset Metadata

Optionally assets may have json metadata attached which can be updated via `GET` / `POST` on 
`/OVEAssetModelController/AssetMeta/{id}`. This is intended to be entered programmatically by asset processing services. 

## Implementation 

> The Asset Manager is implemented in [**C#**](https://github.com/dotnet/roslyn) and [**.net core 2.1**](https://blogs.msdn.microsoft.com/dotnet/2018/05/30/announcing-net-core-2-1/) and so runs **cross platform**. [**OWIN** (Open Web Interface for .Net)](http://owin.org/) is used to decouple the web stack modules with **Dependency Injection** used throughout, precise configuration can be explored in `Program.cs` and `Startup.cs`. Module imports are completed via [**Nuget**](www.nuget.org) and are listed in the `.csproj` files. The cross platform lightweight [**Kestrel**](https://github.com/aspnet/KestrelHttpServer) Http server is used. The opensource [**ASP.Net**](https://github.com/aspnet/AspNetCore) framework is used to create Restful API's, further the [**Asp.net MVC**](https://github.com/aspnet/Mvc) framework and is used to coordinate views, controllers and model state. Model state validation is enabled via a [**Validation Attributes**](https://docs.microsoft.com/en-us/aspnet/core/mvc/models/validation?view=aspnetcore-2.1) and is provided serverside and automatically provided clientside with [**JQuery**](https://jquery.com/). The [**Entity Framework Core**](https://github.com/aspnet/EntityFrameworkCore) framework is used to manage models. [**Code first migrations and deployment**](https://docs.microsoft.com/en-us/aspnet/mvc/overview/getting-started/getting-started-with-ef-using-mvc/migrations-and-deployment-with-the-entity-framework-in-an-asp-net-mvc-application) together with [Scaffolding](https://docs.microsoft.com/en-us/aspnet/mvc/overview/older-versions/hands-on-labs/aspnet-mvc-4-entity-framework-scaffolding-and-migrations) were initially used to generate CRUD templates. Initialisation, maintainence and updates to database structure are managed via database [**Migrations**](https://docs.microsoft.com/en-us/aspnet/mvc/overview/getting-started/getting-started-with-ef-using-mvc/migrations-and-deployment-with-the-entity-framework-in-an-asp-net-mvc-application) which enable automatic in-place database upgrades. [**Razor**](https://docs.microsoft.com/en-us/aspnet/core/mvc/views/razor?view=aspnetcore-2.1) is used for View generation. Styling is acheived via [**Bootstrap**](https://getbootstrap.com/docs/4.1/getting-started/introduction/). Internal processing microservices are implemented using the [**HostedService**](https://blogs.msdn.microsoft.com/cesardelatorre/2017/11/18/implementing-background-tasks-in-microservices-with-ihostedservice-and-the-backgroundservice-class-net-core-2-x/) pattern. A variety of [database adaptors](https://docs.microsoft.com/en-us/ef/core/providers/) are available for EntityFramwork and can be swapped out easily if required, provided currency restrictions are respected. Currently the Asset Manager uses the [**MariaDB**](https://www.nuget.org/packages/Pomelo.EntityFrameworkCore.MySql) adaptor and the [**Amazon S3**](https://www.nuget.org/packages/Amazon.S3/) driver. Full API documentation is acheived using code [xml documention](https://docs.microsoft.com/en-us/dotnet/csharp/codedoc) and [**Swashbuckle**](https://github.com/domaindrivendev/Swashbuckle) to generate [**Swagger**](https://swagger.io/) documenation and ui which can be viewed on `/api-docs/`. 

Author: Will modify this paragraph to make ordering clearer but keep content. 
The level of documentation reflects that most OVE Developers are/will be unfamiliar with the language and ecosystem ;-)

> The Asset Manager is implemented in [**C#**](https://github.com/dotnet/roslyn) and [**.net core 2.1**](https://blogs.msdn.microsoft.com/dotnet/2018/05/30/announcing-net-core-2-1/) and so runs **cross platform**. [**OWIN** (Open Web Interface for .Net)](http://owin.org/) is used to decouple the web stack modules with **Dependency Injection** used throughout, precise configuration can be explored in `Program.cs` and `Startup.cs`. Module imports are completed via [**Nuget**](www.nuget.org) and are listed in the `.csproj` files. The cross platform lightweight [**Kestrel**](https://github.com/aspnet/KestrelHttpServer) Http server is used. The opensource [**ASP.Net**](https://github.com/aspnet/AspNetCore) framework is used to create Restful API's, further the [**Asp.net MVC**](https://github.com/aspnet/Mvc) framework and is used to coordinate views, controllers and model state. Model state validation is enabled via a [**Validation Attributes**](https://docs.microsoft.com/en-us/aspnet/core/mvc/models/validation?view=aspnetcore-2.1) and is provided serverside and automatically provided clientside with [**JQuery**](https://jquery.com/). The [**Entity Framework Core**](https://github.com/aspnet/EntityFrameworkCore) framework is used to manage models. [**Code first migrations and deployment**](https://docs.microsoft.com/en-us/aspnet/mvc/overview/getting-started/getting-started-with-ef-using-mvc/migrations-and-deployment-with-the-entity-framework-in-an-asp-net-mvc-application) together with [Scaffolding](https://docs.microsoft.com/en-us/aspnet/mvc/overview/older-versions/hands-on-labs/aspnet-mvc-4-entity-framework-scaffolding-and-migrations) were initially used to generate CRUD templates. Initialisation, maintainence and updates to database structure are managed via database [**Migrations**](https://docs.microsoft.com/en-us/aspnet/mvc/overview/getting-started/getting-started-with-ef-using-mvc/migrations-and-deployment-with-the-entity-framework-in-an-asp-net-mvc-application) which enable automatic in-place database upgrades. [**Razor**](https://docs.microsoft.com/en-us/aspnet/core/mvc/views/razor?view=aspnetcore-2.1) is used for View generation. Styling is acheived via [**Bootstrap**](https://getbootstrap.com/docs/4.1/getting-started/introduction/). Internal processing microservices are implemented using the [**HostedService**](https://blogs.msdn.microsoft.com/cesardelatorre/2017/11/18/implementing-background-tasks-in-microservices-with-ihostedservice-and-the-backgroundservice-class-net-core-2-x/) pattern. A variety of [database adaptors](https://docs.microsoft.com/en-us/ef/core/providers/) are available for EntityFramwork and can be swapped out easily if required, provided currency restrictions are respected. Currently the Asset Manager uses the [**MariaDB**](https://www.nuget.org/packages/Pomelo.EntityFrameworkCore.MySql) adaptor and the [**Amazon S3**](https://www.nuget.org/packages/Amazon.S3/) driver. Full API documentation is acheived using code [xml documention](https://docs.microsoft.com/en-us/dotnet/csharp/codedoc) and [**Swashbuckle**](https://github.com/domaindrivendev/Swashbuckle) to generate [**Swagger**](https://swagger.io/) documenation and ui which can be viewed on `/api-docs/`. 

Reviewer 2: @davidbirchwork, I think you have not read the above comment carefully. Some points on the content are not along the lines to familiarity with code or not, but inconsistent use of casing. And, also you have not answered the question on _**currency** restrictions_. Did you mean _**current** restrictions_ or _**concurrency** restrictions_?

> The Asset Manager is implemented in [**C#**](https://github.com/dotnet/roslyn) and [**.net core 2.1**](https://blogs.msdn.microsoft.com/dotnet/2018/05/30/announcing-net-core-2-1/) and so runs **cross platform**. [**OWIN** (Open Web Interface for .Net)](http://owin.org/) is used to decouple the web stack modules with **Dependency Injection** used throughout, precise configuration can be explored in `Program.cs` and `Startup.cs`. Module imports are completed via [**Nuget**](www.nuget.org) and are listed in the `.csproj` files. The cross platform lightweight [**Kestrel**](https://github.com/aspnet/KestrelHttpServer) Http server is used. The opensource [**ASP.Net**](https://github.com/aspnet/AspNetCore) framework is used to create Restful API's, further the [**Asp.net MVC**](https://github.com/aspnet/Mvc) framework and is used to coordinate views, controllers and model state. Model state validation is enabled via a [**Validation Attributes**](https://docs.microsoft.com/en-us/aspnet/core/mvc/models/validation?view=aspnetcore-2.1) and is provided serverside and automatically provided clientside with [**JQuery**](https://jquery.com/). The [**Entity Framework Core**](https://github.com/aspnet/EntityFrameworkCore) framework is used to manage models. [**Code first migrations and deployment**](https://docs.microsoft.com/en-us/aspnet/mvc/overview/getting-started/getting-started-with-ef-using-mvc/migrations-and-deployment-with-the-entity-framework-in-an-asp-net-mvc-application) together with [Scaffolding](https://docs.microsoft.com/en-us/aspnet/mvc/overview/older-versions/hands-on-labs/aspnet-mvc-4-entity-framework-scaffolding-and-migrations) were initially used to generate CRUD templates. Initialisation, maintainence and updates to database structure are managed via database [**Migrations**](https://docs.microsoft.com/en-us/aspnet/mvc/overview/getting-started/getting-started-with-ef-using-mvc/migrations-and-deployment-with-the-entity-framework-in-an-asp-net-mvc-application) which enable automatic in-place database upgrades. [**Razor**](https://docs.microsoft.com/en-us/aspnet/core/mvc/views/razor?view=aspnetcore-2.1) is used for View generation. Styling is acheived via [**Bootstrap**](https://getbootstrap.com/docs/4.1/getting-started/introduction/). Internal processing microservices are implemented using the [**HostedService**](https://blogs.msdn.microsoft.com/cesardelatorre/2017/11/18/implementing-background-tasks-in-microservices-with-ihostedservice-and-the-backgroundservice-class-net-core-2-x/) pattern. A variety of [database adaptors](https://docs.microsoft.com/en-us/ef/core/providers/) are available for EntityFramwork and can be swapped out easily if required, provided currency restrictions are respected. Currently the Asset Manager uses the [**MariaDB**](https://www.nuget.org/packages/Pomelo.EntityFrameworkCore.MySql) adaptor and the [**Amazon S3**](https://www.nuget.org/packages/Amazon.S3/) driver. Full API documentation is acheived using code [xml documention](https://docs.microsoft.com/en-us/dotnet/csharp/codedoc) and [**Swashbuckle**](https://github.com/domaindrivendev/Swashbuckle) to generate [**Swagger**](https://swagger.io/) documenation and ui which can be viewed on `/api-docs/`. 

Author: fixed

> **Asset processing microservices may be implemented in any language and interact with the asset manager using its API's**. 

Reviewer: its API's -> its API
